feat: Add self-reschedule guest availability-awareness by host

ImBIOS opened the pull request:
## Todo List

### 1. **Research & Understanding**
   - [x] **Review the current rescheduling logic:** Understand how the system allows hosts to reschedule without checking the guest’s availability.
   - [x] **Identify where guest availability is stored:** Confirm how and where guest availability is stored or accessed in the system.
   - [x] (prisma or trpc I guess) **Understand user identification:** Investigate how the system differentiates between Cal.com users and non-users based on their email.

### 2. **Feature Development**
   - [x] **Check if the guest is a Cal.com user:**
     - [x] Implement logic to verify if the guest (email) is a registered Cal.com user.
   - [x] **Retrieve guest availability:**
     - [x] (auto) If the guest is a registered user, fetch their available times.
   - [x] **Update rescheduling options:**
     - [x] (auto) Modify the rescheduling UI to only show time slots that match both the host's and guest's availability.
     - [x] Ensure that the rescheduling process remains unchanged if the guest is not a registered user.

### 3. **Testing**
   - [x] **E2E Tests:**
     - [x] Test the entire rescheduling process to ensure it accounts for guest availability correctly.
     - [x] Test scenarios with both registered and unregistered guests.
   - [x] (auto) **UI/UX Testing:**
     - [x] Ensure that the UI updates correctly and is intuitive for users when rescheduling.

### 4. **Documentation**
   - [ ] **Update the system documentation:**
     - [ ] Document the new rescheduling process, including the logic for checking guest availability.
     - [ ] Provide usage examples for users.

## What does this PR do?

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #16378 
- /claim #16378

<!-- Please provide a loom video for visual changes to speed up reviews
 Loom Video: https://www.loom.com/
-->

## Mandatory Tasks (DO NOT REMOVE)

- [x] I have self-reviewed the code (A decent size PR without self-review might be rejected).
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com/docs). If N/A, write N/A here and check the checkbox.
- [x] I confirm automated tests are in place that prove my fix is effective or that my feature works.

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set? none
- What are the minimal test data to have? none
- What is expected (happy path) to have (input and output)? idk what it means
- Any other important info that could help to test that PR: `yarn e2e reschedule.e2e.ts`

ImBIOS committed:
feat: Add guest availability awareness for host self reschedule

ImBIOS committed:
chore: Remove commented out code for guest availability-awareness reschedule

ImBIOS committed:
refactor: Remove commented out code for guest availability-awareness reschedule

ImBIOS committed:
test: Implement initial e2e tests

ImBIOS committed:
Merge branch 'main' into feat/reschedule-guest-availability-awareness

ImBIOS committed:
refactor: Remove unused code for CRM contact owner retrieval

ImBIOS committed:
feat: Add logic to retrieve booker's calendar credentials and availability

ImBIOS commented:
LOL :rofl:, I wonder why accidentally my solution broken. It seems someone just changed how the useSchedule work. I'm working on it again now. Please be patient.

ImBIOS committed:
fix: Reintroduce bookerEmail

ImBIOS committed:
Merge branch 'main' into feat/reschedule-guest-availability-awareness

ImBIOS commented:
Whoever reviews it, the `turbo.json` and `i18n.json` are just prettier runs when I execute git merge and resolve conflict. It seems someone forgot to activate prettier and it's somehow merged to the main branch. You can ignore those 2 files.

ImBIOS commented:
Should I document the reschedule feature here: https://github.com/calcom/docs?

Amit91848 commented on packages/features/bookings/Booker/utils/event.ts:
```diff
@@ -108,6 +110,7 @@ export const useScheduleForEvent = ({
     isTeamEvent: pathname?.indexOf("/team/") !== -1 || isTeam,
     orgSlug,
     teamMemberEmail,
+    bookerEmail,
```
This was recently removed because it was calling getSchedules endpoint at every key stroke unnecessarily. I am not sure if you added this intentionally or was some merge issue, please remove this.


https://github.com/user-attachments/assets/c8cd2efc-8bd5-453f-af92-550469a8c56a

Amit91848 commented on packages/features/schedules/lib/use-schedule/useSchedule.ts:
```diff
@@ -34,6 +35,7 @@ export const useSchedule = ({
   isTeamEvent,
   orgSlug,
   teamMemberEmail,
+  bookerEmail,
```
Same here

Amit91848 commented on packages/platform/atoms/booker/BookerWebWrapper.tsx:
```diff
@@ -131,6 +131,7 @@ export const BookerWebWrapper = (props: BookerWebWrapperAtomProps) => {
     duration: props.duration,
     selectedDate,
     teamMemberEmail: props.teamMemberEmail,
+    bookerEmail: bookerForm.formEmail,
```
Same here

Amit91848 commented on packages/platform/atoms/booker/BookerWebWrapper.tsx:
```diff
@@ -140,6 +141,8 @@ export const BookerWebWrapper = (props: BookerWebWrapperAtomProps) => {
     teamMemberEmail: props.teamMemberEmail,
   });
 
+  console.log("bookerForm", bookerForm);
```
NIT: remove logs

Amit91848 commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -319,9 +319,22 @@ export interface IGetAvailableSlots {
+        },
+      })
+    : null;
+  console.log("input", input);
```
Same here

Amit91848 commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -319,9 +319,22 @@ export interface IGetAvailableSlots {
 }
 
 export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Promise<IGetAvailableSlots> {
+  const bookerCalUser = input.bookerEmail
```
In case of collective events, or if you have added additional guests, can you check their availabilities as well?

Amit91848 reviewed (changes requested):
Thanks for the PR @ImBIOS , works well, but I have some concerns and comments. Lmk what you think.

Also type check seems to be failing, can you check that?

ImBIOS committed:
feat: Remove console.log statements and improve code readability

ImBIOS committed:
feat: Update useScheduleForEvent to use attendeeEmails instead of bookerEmail

ImBIOS committed:
feat: Refactor getAvailableSlots function to include attendee calendar users

ImBIOS commented:
@Amit91848 Currently it's ready to review again, and I have a question regarding `isFixed` in the `util.ts`, I don't know what this data is for, I set the default for all attendees top `false`.

Amit91848 commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -413,15 +427,17 @@ export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Pro
           return usersArray;
-        }, [] as (GetAvailabilityUser & { isFixed: boolean })[]);
+        }, [] as (GetAvailabilityUser & { isFixed: boolean })[])),
+    ...(attendeeCalUsers ? attendeeCalUsers.map((user) => ({ ...user, isFixed: false })) : []),
```
> @Amit91848 Currently it's ready to review again, and I have a question regarding isFixed in the util.ts, I don't know what this data is for, I set the default for all attendees top false.


For users who have isFixed: true (collective event hosts and if user adds fixed hosts for round robin events) they **all have to available** at the given time slot. It is optional for non fixed users (round robin hosts where only one of hosts needs to be present)

So in this case we want attendee users to also be present, isFixed here should be true
```suggestion
    ...(attendeeCalUsers ? attendeeCalUsers.map((user) => ({ ...user, isFixed: true })) : []),
```

Amit91848 committed:
Merge branch 'main' into feat/reschedule-guest-availability-awareness

ImBIOS committed:
fix: Set attendeeCalUsers isFixed default value to true

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

ImBIOS committed:
fix: Fix type error

ImBIOS committed:
feat: Update reschedule.e2e.ts to set default name for guests without username

ImBIOS committed:
feat: Refactor setGuestWorkingHours12PM5PM function

The `setGuestWorkingHours12PM5PM` function in `reschedule.e2e.ts` has been refactored to improve code readability and remove unnecessary imports.

ImBIOS commented:
@Amit91848 Thank God, the type-check seems success now.

Amit91848 commented on apps/web/playwright/reschedule.e2e.ts:
```diff
@@ -416,6 +418,112 @@ test.describe("Reschedule Tests", async () => {
+        },
+      });
+
+      await setGuestWorkingHours12PM5PM({ page, host, guest: guests[0] });
```
Can it be done via user.create fixture instead? Not really a big fan of using e2e to do it, they are very flaky plus time consuming.

ImBIOS replied:
nope, currently we can't do it using user fixture

Amit91848 commented on apps/web/playwright/reschedule.e2e.ts:
```diff
@@ -416,6 +418,112 @@ test.describe("Reschedule Tests", async () => {
+        },
+      });
+
+      await page.goto(`/reschedule/${booking.uid}`);
```
```suggestion
      await page.goto(`/reschedule/${booking.uid}`);
      await page.waitForLoadState("networkidle");
```

Please wait for page loads to complete wherever necessary, or else it causes pages to be flaky. Might work locally but when run on ci, it tends to fail sometimes.

Amit91848 commented on apps/web/playwright/reschedule.e2e.ts:
```diff
@@ -416,6 +418,112 @@ test.describe("Reschedule Tests", async () => {
+    }) => {
+      const host = await users.create();
+      const guests = await Promise.all(
+        Array.from({ length: 10 }, (_, i) => users.create({ username: `guest-${i}` }))
```
No need to add these many guests. Add 2 guests with separate availability timings and having some bookings for them.

Amit91848 reviewed (changes requested):
Thanks for such quick updates @ImBIOS . Suggested some improvements in e2e flow. 
Please avoid using e2e for stuff outside of tests instead try updating the fixtures first. E2E is very flaky and can have multiple points of failure making it hard to debug and fix all.

Amit91848 commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -319,9 +319,23 @@ export interface IGetAvailableSlots {
 }
 
 export async function getAvailableSlots({ input, ctx }: GetScheduleOptions): Promise<IGetAvailableSlots> {
+  const attendeeCalUsers = input.attendeeEmails
```
We are already fetching original booking details in line 413. You can also get attendees from there instead of passing it via frontend.

ImBIOS replied:
you're right, thanks for noticing it.

ImBIOS commented on packages/trpc/server/routers/viewer/slots/util.ts:
```diff
@@ -319,9 +319,23 @@ export interface IGetAvailableSlots {
       emoji?: string | undefined;
     }[]
   >;
+  teamMember?: string | undefined;
```
it's unintentional, I think I accidentally add it while fixing merge conflict. removing it...

ImBIOS committed:
test(reschedule): Add wait for network idle

It's to make sure the page is fully hydrated to reduce test flakiness

Co-authored-by: Amit Sharma <74371312+Amit91848@users.noreply.github.com>

ImBIOS committed:
feat: Update reschedule.e2e.ts to set default name for guests without username

ImBIOS committed:
Merge branch 'main' into feat/reschedule-guest-availability-awareness

ImBIOS committed:
feat: Add rescheduling with multiple attendees test cases

ImBIOS committed:
test: Improve rescheduling test cases for guest availability awareness

Amit91848 commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -477,6 +478,25 @@ async function handler(
+      eventTypeId: eventTypeId,
+      eventTypeSlug: eventTypeSlug,
+      timeZone: reqBody.timeZone,
+      usernameList: dynamicUserList,
```
Why did you use dynamicUserList? It only has users which are added as users in the url for dynamic events?

Amit91848 commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -477,6 +478,25 @@ async function handler(
+      eventTypeSlug: eventTypeSlug,
+      timeZone: reqBody.timeZone,
+      usernameList: dynamicUserList,
+      duration: reqEventLength - 1,
```
Why is - 1 required?

Amit91848 commented on packages/features/bookings/lib/handleNewBooking.ts:
```diff
@@ -477,6 +478,25 @@ async function handler(
     }
   }
 
+  const availableSlots = await getAvailableSlots({
```
getAvailableSlots and later on ensureAvailableUsers both call getUsersAvailability internally. 
Can you confirm it does not adversely affect the performance of this endpoint? 

You should also compare it with doing it by adding guests to users that are passed to ensureUsersAvailabilityo

Amit91848 reviewed (changes requested):
Great work @ImBIOS . I am concerned about performance when using `getAvailableSlots` inside 
`handleNewBooking`. Lmk your thoughts.
Also I see you have added unit tests which is great. You should remove the e2e changes added as unit tests should be good enough 👍

Amit91848 reviewed (changes requested):
We still want to allow the guests themselves to reschedule outside of their availability. So we only want to check for the guest's availability when the event type owners (or team admins) reschedule.

See https://github.com/calcom/cal.com/issues/16378#issuecomment-2347342541

anikdhabal commented:
@ImBIOS closing it as we found the other one is close to solution and faster
